### PR TITLE
ENH: stats: add `alternative` to masked version of T-Tests

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
#### Reference issue

Addresses gh-12506

#### What does this implement/fix?

Masked versions of T-Tests were missing an `alternative` parameter.
This adds the parameter to `ttest_1samp`, `ttest_ind`, and `ttest_rel`.
As far as I know, this should cover all the remaining T-Tests missing the
`alternative` parameter.

#### Additional information

I am willing to add descriptive documentation of the alternative parameter and a few examples in the documentation. What do you think, @mdhaber? Marking this as draft while I work on it. Otherwise, I think, this should be ready for review

### CODE REVIEW

**Comment by mdhaber** [comment]
Cool. 

- Can you add in description of the alternatives? (We should have done this for `stats.ttest_ind`, but that hadn't yet occurred to me when we did that.)
- Would it be much trouble to share `_ttest_finish` between `stats` and `mstats`?
- Is it convenient to share (maybe copy-paste and indicate with a comment, maybe import) some tests between mstats and stats?

IMO these codes shouldn't be separate, but I'm not asking you to fix that - just asking you to see if we can minimize the divergence added in this PR.

File: scipy/stats/tests/test_mstats_basic.py [hunk 1]
```diff
@@ -1085,6 +1085,24 @@ def test_zero_division(self):
             assert_array_equal(t, np.array([np.nan, np.nan]))
             assert_array_equal(p, np.array([np.nan, np.nan]))
 
+    def test_alternative(self):
```
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:1088]: I think these `mstats_basic` tests are fine. Ideally, if this test is based on an existing two-sided test that compares `mstats` with `stats`, you could just parametrize that, but this is OK too.

**Comment by tirthasheshpatel** [comment]
>  Can you add in description of the alternatives? (We should have done this for `stats.ttest_ind`, but that hadn't yet occurred to me when we did that.)

On it. That's why this is a draft for now.

>  Would it be much trouble to share `_ttest_finish` between `stats` and `mstats`?

I think so. `mstats` uses `stats.betainc` to calculate the CDF of the T distribution as it handles masked inputs without warnings. `stats.t.cdf` doesn't handle masked inputs and throws `RuntimeWarning`s. So, I think, it is best to keep the two functions apart. If you want, I can change the name to something more apparent like `_masked_ttest_finish`.

> IMO these codes shouldn't be separate

I agree. But there are already a lot of tests with such disparity. Do you think, in the future, it would be worthwhile to deprecate the separate masked versions and merge them with the tests in `stats.py`. If so, it would be good to create an issue and see others' thoughts.

File: scipy/stats/tests/test_mstats_basic.py (hunk 1 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_mstats_basic.py:1088]: > Ideally, if this test is based on an existing two-sided test that compares `mstats` with `stats`, you could just parametrize that to go through all the alternatives, but this is OK too.

Makes sense. I will try to do that.

**Comment by mdhaber** [comment]
> Do you think, in the future, it would be worthwhile to deprecate the separate masked versions and merge them with the tests in stats.py

I don't know if we can really deprecate them, but I do think we should try to unify the implementations. It's not a priority for me right now, though. Lots of work for not much gain.

**Comment by tirthasheshpatel** [comment]
I have added descriptive docs for (I think) all the remaining T-tests. I haven't copy-pasted that to the masked version as it already points to the main documentation.

**Comment by mdhaber** [comment]
> I think so. `mstats` uses `stats.betainc` to calculate the CDF of the T distribution as it handles masked inputs without warnings.

Well why don't we use that in `stats._ttest_finish`?

How were we able to use `scipy.stats.stats._normtest_finish` in gh-13960?

File: scipy/stats/tests/test_mstats_basic.py [hunk 2]
```diff
@@ -1083,6 +1083,21 @@ def test_zero_division(self):
             assert_array_equal(t, np.array([np.nan, np.nan]))
             assert_array_equal(p, np.array([np.nan, np.nan]))
 
+    def test_bad_alternative(self):
+        msg = r"`alternative` must be 'less', 'greater' or 'two-sided'"
+        with pytest.raises(ValueError, match=msg):
+            mstats.ttest_ind([1, 2, 3], [4, 5, 6], alternative='foo')
+
+    @pytest.mark.parametrize("alternative", ["less", "greater"])
```
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:1091]: Why is this parametrized and the tests below are not? Why is `test_bad_alternative` a separate function above but not below?

File: scipy/stats/tests/test_stats.py [hunk 3]
```diff
@@ -2881,6 +2881,22 @@ def test_1samp_alternative(self):
         assert_allclose(p, self.P1_1_g)
         assert_allclose(t, self.T1_1)
 
+    def test_1samp_nan_alternative(self):
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:2884]: These tests fit into the existing tests. I don't like the inconsistencies of the existing tests (e.g. some functions test 1d, some 2d, some 3d, and only ever on one axis), but fixing that is out of scope.

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:2884]: I seems that the purpose of this test is to be an intuitive one. But why have an intuitive test for `ttest_1samp` and not the others?

**Comment by tirthasheshpatel** [comment]
> Well why don't we use that in `stats._ttest_finish`? I might have prefered `stats.t.cdf` for readability, but with a comment that `stats.betainc` it should be fine, and it would avoid the overhead, too.

Oh, of course! We can do that.

> How were we able to use `scipy.stats.stats._normtest_finish` in [gh-13960](https://github.com/scipy/scipy/pull/13960)?

I think `norm.cdf` handles masked arrays without any problems while `t.cdf` throws some warnings. So, we were able to directly use the `_normtest_finish` in gh-13960

File: scipy/stats/tests/test_stats.py (hunk 3 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:2884]: > These tests fit into the existing tests. I don't like the inconsistencies of the existing tests (e.g. some functions test 1d, some 2d, some 3d, and only ever on one axis), but fixing that is out of scope.

I agree. We can fix that in another PR. Do you think tracking this in an issue would be helpful?

> I seems that the purpose of this test is to be an intuitive one. But why have an intuitive test for `ttest_1samp` and not the others?

I might have added this by mistake. Didn't mean to add an intuitive test here. I didn't add intuitive tests because it has already been done while adding `alternative` parameter. Also, tests against R have been added there. So, I will remove this.

File: scipy/stats/tests/test_mstats_basic.py (hunk 2 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_mstats_basic.py:1091]: > Why is this parametrized and the tests below are not? Why is `test_bad_alternative` a separate function above but not below?

Both are mistakes. I will parameterize and separate the two tests like I have done here.

> If we're testing `mstats`, shouldn't we test with a masked array with masked elements?

Yes, we should. I will add these in the next commit.

**Comment by tirthasheshpatel** [comment]
@mdhaber I have tried to resolve the refguide failures. So, this should be ready for another review.

File: scipy/stats/tests/test_mstats_basic.py [hunk 4]
```diff
@@ -1124,6 +1124,32 @@ def test_zero_division(self):
...
 
+    def test_bad_alternative(self):
+        msg = r"alternative must be 'less', 'greater' or 'two-sided'"
+        with pytest.raises(ValueError, match=msg):
+            mstats.ttest_ind([1, 2, 3], [4, 5, 6], alternative='foo')
+
+    @pytest.mark.parametrize("alternative", ["less", "greater"])
+    def test_alternative(self, alternative):
+        x = stats.norm.rvs(loc=10, scale=2, size=100, random_state=123)
+        y = stats.norm.rvs(loc=8, scale=2, size=100, random_state=123)
```
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:1135]: I would suggest adjusting the way the random state is fixed. Doing it this way, every element of x is bigger than the corresponding element of y, so we get p-values of either 0 or 1. That also means there's no distinction in the p-value for `alternative="greater"` vs `alternative='two-sided'`. The test would be a little more sensitive if it had to calculate the correct decimal value, which would be distinct for each alternative.

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -1469,7 +1469,7 @@ def test_regress_alternative(self):
 
         # slope is greater than zero, so "greater" p-value should be small
         res3 = stats.linregress(x, y, alternative="greater")
-        assert_allclose(res3.pvalue, res1.pvalue / 2)
+        assert_allclose(res3.pvalue, res1.pvalue / 2, atol=1e-16)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:1472]: The p-value is `2.072831043328079e-92`, so if there is an `atol` here, the test is not very sensitive.

This is sort of like the issue in the `ttest_rel` test above. We should have tested p-values that are a little less extreme.

Was this change necessary?
 (probably not now that we're implementing `_ttest_finish` with `stdtr`?)

File: scipy/stats/tests/test_stats.py [hunk 6]
```diff
@@ -3793,6 +3794,28 @@ def test_ttest_rel():
...
+
+    rvs1_2D[:, 20:30] = np.nan
+    rvs2_2D[:, 15:25] = np.nan
+
+    tr, pr = stats.ttest_rel(rvs1_2D, rvs2_2D, 0, nan_policy='omit')
+
+    t, p = stats.ttest_rel(rvs1_2D, rvs2_2D, 0, nan_policy='omit',
+                           alternative='less')
+    assert_allclose(t, tr, atol=1e-12)
+    assert_allclose(p, converter(tr, pr, 'less'), atol=1e-12)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:3812]: Why `atol` here? If you're trying to make the tolerance more strict, use `rtol`. 
If you need it to be looser, why? If it can't pass `rtol`, we should probably not have a test with such small p-values (because with `atol` you can lose the distinction between the two-sided and one-sided p-values).

File: scipy/stats/tests/test_stats.py [hunk 7]
```diff
@@ -3793,6 +3794,28 @@ def test_ttest_rel():
...
 
+    # Convert from two-sided p-values to one sided using T result data.
+    def convert(t, p, alt):
+        if (t < 0 and alt == "less") or (t > 0 and alt == "greater"):
+            return p / 2
+        return 1 - (p / 2)
+    converter = np.vectorize(convert)
+
+    rvs1_2D[:, 20:30] = np.nan
+    rvs2_2D[:, 15:25] = np.nan
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:3805]: These are fine, but thoughts for the future.
* randomize the `nan` patterns
* test along each axis (`@parametrize`)
* test more than two dimensional arrays. I've seen functions that work for two dimensions, but not three or more.
* since `ttest_rel` without `nan`s is already tested for accuracy, test against that (i.e. loop over it, manually stripping out the `nan`s from each axis-slice)

Let's not change these, though. What is here is pretty consistent with the rest, and we need some infrastructure (e.g. gh-13312 ) to make some of these suggestions easy to implement.

**Review by mdhaber — COMMENTED** [review]
Just a few last things.

File: scipy/stats/mstats_basic.py [hunk 8]
```diff
@@ -979,7 +979,7 @@ def sen_seasonal_slopes(x):
 Ttest_1sampResult = namedtuple('Ttest_1sampResult', ('statistic', 'pvalue'))
 
 
-def ttest_1samp(a, popmean, axis=0):
+def ttest_1samp(a, popmean, axis=0, alternative='two-sided'):
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply scipy/stats/mstats_basic.py:982]: `mean -> means` change done. Thanks!

File: scipy/stats/tests/test_mstats_basic.py (hunk 4 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_mstats_basic.py]: I changed the scale of x to 5 and decreased the sample size to 25 with seed=42. This now generates non-zero p-values which would test ttest_rel alternatives up to a few decimal points.

File: scipy/stats/tests/test_stats.py (hunk 5 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py]: > Was this change necessary?
> (probably not now that we're implementing `_ttest_finish` with `stdtr`?)

No, this was only necessary when we were using `betainc` in `_ttest_finish`. With `betainc -> stdtr` change, we can revert this.

File: scipy/stats/tests/test_stats.py (hunk 6 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py]: > Why `atol` here? If you're trying to make the tolerance more strict, use `rtol`.

Thanks for the suggestion. Done.



> If you need it to be looser, why? If it can't pass `rtol`, we should probably not have a test with such small p-values (because with `atol` you can lose the distinction between the two-sided and one-sided p-values).

Oh, didn't know that atol loosens the test. I will keep this in mind from now on :)

File: scipy/stats/tests/test_stats.py (hunk 7 above)
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:3805]: > These are fine, but thoughts for the future.
> 
> * randomize the `nan` patterns
> 
> * test along each axis (`@parametrize`)
> 
> * test more than two dimensional arrays. I've seen functions that work for two dimensions, but not three or more.
> 
> * since `ttest_rel` without `nan`s is already tested for accuracy, test against that (i.e. loop over it, manually stripping out the `nan`s from each axis-slice)

OK. All these suggestions sound good. I will do these from now on.

File: scipy/stats/tests/test_stats.py (hunk 6 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py]: If the reference p-value were exactly zero, then it would be impossible to get the desired relative tolerance unless the actual result were exactly zero. Adding an absolute tolerance, then, would effectively loosen the tolerance. That's all I meant.

File: scipy/stats/tests/test_stats.py (hunk 7 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:3805]: They might be overkill. And until gh-gh-13312 gives an easy way of creating a reference implementation of `nan_policy` over multiple dimensions, it's not worth the trouble. So just thinking ahead.

**Review by mdhaber — APPROVED** [review]
I am rerunning a build that I think failed temporarily. If that passes, I'll go ahead and merge.

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel!
